File: src/python/transaction/working_with_transaction/pushing_complex_operation/complex_operation.py
import asyncio

import wax
from wax import create_hive_chain
from wax.proto.operations import witness_set_properties
from wax.wax_result import python_witness_set_properties_data

# Initialize chain
chain = create_hive_chain()


async def main() -> None:
    # Initialize a transaction object
    trx = await chain.create_transaction()

    wsp_operation = witness_set_properties(
        owner="alice",
        props=wax.serialize_witness_set_properties(
            python_witness_set_properties_data(
                url=b"https://example.com", key="STM57gC3aqyDvu2fPPdfpY2iDtLU6PDb8qD8RGmfxLf1q43PhJYYQ"
            )
        ),
    )

    trx.push_operation(wsp_operation)

    # Get a transaction object holding all operations and transaction TAPOS & expiration data, but transaction is **not signed yet**
    built_transaction = trx.transaction
    print(built_transaction.operations[0])  # Witness set properties operation


asyncio.run(main())

File: src/rust/transaction/working_with_transaction/pushing_complex_operation/complex_operation.rs
use runner::{SnippetsBeekeeperData, snippets_beekeeper_data};
use wax::prelude::*;

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    // Initialize chain
    let chain = create_hive_chain(None)?;

    let SnippetsBeekeeperData { public_key1, .. } = snippets_beekeeper_data();

    // Initialize a transaction object
    let mut tx = chain.create_transaction(None).await?;

    // Build operation
    tx.push_complex_operation(
        &chain,
        WitnessSetPropertiesOperation {
            owner: "owner".into(),
            witness_signing_key: public_key1,
            url: Some("https://example.com".into()),
            ..Default::default()
        },
    )?;

    /*
    Get a transaction object holding all operations and transaction
    TAPOS & expiration data, but transaction is **not signed yet**
    */
    let built_transaction = tx.transaction();

    // Witness set properties operation
    println!("{:#?}", built_transaction.operations[0]);

    Ok(())
}
